### PR TITLE
R&D Monitoring console TGUI + Can see RD consoles

### PR DESCRIPTION
**Pull request by JohnFulpWillard**
## About The Pull Request

I wrote this while constantly rushing lol

This PR does many things, the largest is that the R&D Monitoring console (the RD's one) is now TGUI
It also changes how researching loggings work, bringing the RD console and NtosRD app on par with eachother

The type of person is also logged differently, instead of ``Cyborg: [name]`` and ``User: [name]``, humans do not have a prefix and cyborgs will say ``CYBORG [name]``. ``research_logs`` also works differently now to make it easier to reference each needed data from the list.

Lastly, I added the ability to see R&D consoles from the console, and the ability to remotely un/lock them down. This currently is pretty useless as it can't control the tablet app variant, and anyone with Science access can just unlock it, however with some minor future changes I think this can be turned into a good way for the RD to get control of their department.

Video demonstration, mostly (I made a few edits after this):
https://user-images.githubusercontent.com/53777086/215005387-817106f4-5237-4f2e-b0ac-da28e6a17f9c.mp4

## Why It's Good For The Game

This console is overhyped by the game, being hidden behind an RD-locked Command-colored door, in the same room as one of the most damaging theft objectives, yet it is one of the most useless and forgotten consoles in R&D if you don't count everything outside of researching, experiments and robotics.

This adds a nice TGUI menu while making it a little more worthwhile to use.

## Changelog

:cl:
balance: The R&D monitoring console now shows R&D consoles and their locations.
refactor: The R&D monitoring console now has a nice TGUI menu.
/:cl:

### CODE REVIEW

**Comment by jlsnow301** [comment]
good work, i think this would look really nice if you put the research log in a scrollable section with the elements as classname="candystripe"

also, i don't IMAGINE there's ever a case where there's too many servers and consoles going on, but how would that make the rest of the UI look?

**Comment by moocowswag** [comment]
Don`t console locks just get stopped by a simple screwdriver x2

**Comment by OrionTheFox** [comment]
Juuust to clarify, was the list of research/where its done sorted with Recent on the bottom? Might be best to keep it as most recent at the top, since the current one is that way and scrolling through the full day's research to look at a recent change will get tiring.

**Comment by JohnFulpWillard** [comment]
> good work, i think this would look really nice if you put the research log in a scrollable section with the elements as classname="candystripe"

I did that originally but the UI kept automatically scrolling down to the scrollable section, moving the consoles/servers out of view, even if I scrolled back up

**Comment by JohnFulpWillard** [comment]
> Juuust to clarify, was the list of research/where its done sorted with Recent on the bottom? Might be best to keep it as most recent at the top, since the current one is that way and scrolling through the full day's research to look at a recent change will get tiring.

It doesn't update when something new is researched so it should be fine, I liked it better like this from testing because it's in order of research

**Comment by jlsnow301** [comment]
> I did that originally but the UI kept automatically scrolling down to the scrollable section, moving the consoles/servers out of view, even if I scrolled back up

would you be willing to set it as requested, but without the scrollable prop then? for readability

**Comment by JohnFulpWillard** [comment]
@jlsnow301 done (Opened the UI a bit more to show it off)

image

**Comment by zxaber** [comment]
How is the user determined? Is it based on who logged into the program last?

--

> > Don`t console locks just get stopped by a simple screwdriver x2
> 
> sad to hear, maybe consoles should start locked if accessed by someone who doesn't have access to the console but that's not for this PR

You could add a locked variable to the board, so that rebuilding with a locked board still results in it being locked. Fluffwise, you'd be locking the board's MAC or something.

File: tgui/packages/tgui/interfaces/ServerControl.tsx [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+          </NoticeBox>
+        ) : (
+          <Section>
+            <Table cellpadding="3" textAlign="center">
+              <Table.Row header>
+                <Table.Cell>Research Servers</Table.Cell>
+              </Table.Row>
+              {servers.map((server) => (
+                <>
+                  <Table.Row header key={server} />
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/ServerControl.tsx:63]: candystripe please

File: tgui/packages/tgui/interfaces/ServerControl.tsx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        )}
+
+        {!consoles ? (
+          <NoticeBox mt={2} info>
+            No consoles found.
+          </NoticeBox>
+        ) : (
+          <Section align="right">
+            <Table cellpadding="3" textAlign="center">
+              <Table.Row header>
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/ServerControl.tsx:91]: and here

File: tgui/packages/tgui/interfaces/ServerControl.tsx [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+          </Section>
+        )}
+
+        <Collapsible title="Research History">
+          {!logs.length ? (
+            <NoticeBox mt={2} info>
+              No history found.
+            </NoticeBox>
+          ) : (
+            <Table>
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/ServerControl.tsx:126]: please also wrap this in a section, the white on empty gray is still present

**Comment by JohnFulpWillard** [comment]
> How is the user determined for ModPC app versions? Is it based on who logged into the program last?

It doesn't currently:

> This currently is pretty useless as it can't control the tablet app variant

Locking down RD consoles has always existed so this problem isn't limited to this PR, it just makes it a little more apparent of its existence, so I'll leave it for another PR (or anyone who wants it done a certain way and gets to it before me as I am slow)

File: tgui/packages/tgui/interfaces/ServerControl.tsx (hunk 3 above)
Comment (JohnFulpWillard) [reply tgui/packages/tgui/interfaces/ServerControl.tsx]: image

**Pull request merged**